### ISSUE TITLE
Reduce GPU memory usage?

### ISSUE THREAD

**Issue by MichaelBrambleberry** [ultralytics/ultralytics#21323]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Trying to use YOLOv11n with a python GUI to detect objects on screen in videogames.

I was able to get a solid 15-20 FPS using CPU. I wanted to increase this by using GPU/CUDA, however when I use CUDA, the model.predict() on python causes the program to use ~60-80% of my GPU (NVIDIA RTX A4500). The videogame I am running is already using ~40% of my GPU in task manager, which causes 100% cumulative usage and causes <1 FPS.

I have tried retraining my custom model with different parameters, such as batch=0.20, which Google AI was saying would limit the inference to using 20% of GPU memory.

It seems that no matter what parameters I change, the GPU usage for inference / prediction remains the same, very high. I was able to get decent performance with CPU, I thought that I would be able to get slightly better performance using only a fraction of the GPU, since the GPU is much more powerful. I am fine with slightly slower inference.

What am I doing wrong? Is YOLOv11 not an appropriate tool for this application or is there a setting I am missing to lock the GPU usage properly?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MichaelBrambleberry, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will review your question and assist you soon.

We recommend checking the [Docs](https://docs.ultralytics.com/) for guidance on optimizing YOLO usage and reviewing [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions are addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the issue.

For custom usage or performance ❓ questions like yours, please include as much detail as possible—such as your code snippet for model loading and inference, GPU details, and any logs or screenshots that illustrate the GPU usage you’re seeing. Also, refer to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for more insights.

Join our Ultralytics community for additional support and discussion:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth technical discussion, see [Discourse](https://community.ultralytics.com/)
- To share tips and projects, check our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

To ensure your issue is not already resolved, upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

We appreciate your patience and look forward to helping you resolve your GPU memory usage question!

**Comment by Y-T-G** [comment]
You can slow down the inference, by manually adding `time.sleep()` between each inference call. Otherwise, the model will simply infer on as many frames as possible.

If you have an NVIDIA GPU, then you should also export to TensorRT with `half=True` for better optimization:

https://docs.ultralytics.com/integrations/tensorrt/#usage

**Comment by MichaelBrambleberry** [comment]
> You can slow down the inference, by manually adding `time.sleep()` between each inference call. Otherwise, the model will simply infer on as many frames as possible.
> 
> If you have an NVIDIA GPU, then you should also export to TensorRT with `half=True` for better optimization:
> 
> https://docs.ultralytics.com/integrations/tensorrt/#usage



I've tried time.sleep as well as using a frame counter to only use the inference call on every other frame. This improves performance but every time the inference call actually happens, there is still a massive stutter because the GPU memory usage spikes to 100% (60% from the script, 40% from other programs).

 I need to actually hard-limit the amount of GPU memory that is allowed to be used by YOLO inference. Surely there is a way to only allow ~30% of my GPU to be used for inference?

**Comment by Y-T-G** [comment]
You can limit GPU memory usage:
https://docs.pytorch.org/docs/stable/generated/torch.cuda.set_per_process_memory_fraction.html#torch.cuda.set_per_process_memory_fraction

But limiting GPU memory usage wouldn't limit GPU compute usage which is something else. What you're referring to is GPU compute usage, not memory usage.

**Comment by MichaelBrambleberry** [comment]
Is there a way to specifically limit GPU compute usage? Which is the one that is likely causing crashes when run in parallel with a game? It seems to be whichever metric task manager reports as "GPU: 60%"

**Comment by Y-T-G** [comment]
You can try adding this before your code:

```python
import torch
import time

class GPUOpThrottler:
    def __init__(self, delay=0.01):
        self.delay = delay
        self.original_ops = {}
        self.enabled = False
    
    def throttle_tensor_ops(self):
        """Throttle common GPU tensor operations"""
        if self.enabled:
            return
            
        # List of operations to throttle
        ops_to_throttle = [
            'matmul', 'mm', 'conv2d', 'conv1d', 'conv3d',
            'linear', 'relu', 'sigmoid', 'tanh', 'softmax'
        ]
        
        # Monkey patch torch functions
        for op_name in ops_to_throttle:
            if hasattr(torch, op_name):
                original_op = getattr(torch, op_name)
                self.original_ops[op_name] = original_op
                
                def make_throttled_op(orig_op):
                    def throttled_op(*args, **kwargs):
                        result = orig_op(*args, **kwargs)
                        time.sleep(self.delay)
                        return result
                    return throttled_op
                
                setattr(torch, op_name, make_throttled_op(original_op))
        
        # Monkey patch torch.nn.functional
        import torch.nn.functional as F
        functional_ops = ['conv2d', 'linear', 'relu', 'softmax']
        
        for op_name in functional_ops:
            if hasattr(F, op_name):
                original_op = getattr(F, op_name)
                self.original_ops[f'F.{op_name}'] = original_op
                
                def make_throttled_functional_op(orig_op):
                    def throttled_op(*args, **kwargs):
                        result = orig_op(*args, **kwargs)
                        time.sleep(self.delay)
                        return result
                    return throttled_op
                
                setattr(F, op_name, make_throttled_functional_op(original_op))
        
        self.enabled = True
    
    def remove_patches(self):
        """Remove all monkey patches"""
        if not self.enabled:
            return
            
        import torch.nn.functional as F
        
        for op_name, original_op in self.original_ops.items():
            if op_name.startswith('F.'):
                setattr(F, op_name[2:], original_op)
            else:
                setattr(torch, op_name, original_op)
        
        self.original_ops.clear()
        self.enabled = False

# Usage
throttler = GPUOpThrottler(delay=0.005)
throttler.throttle_tensor_ops()
```

You can increase/decrease the delay for more/less throttling.

**Comment by MichaelBrambleberry** [comment]
> You can try adding this before your code:
> 
> import torch
> import time
> 
> class GPUOpThrottler:
>     def __init__(self, delay=0.01):
>         self.delay = delay
>         self.original_ops = {}
>         self.enabled = False
>     
>     def throttle_tensor_ops(self):
>         """Throttle common GPU tensor operations"""
>         if self.enabled:
>             return
>             
>         # List of operations to throttle
>         ops_to_throttle = [
>             'matmul', 'mm', 'conv2d', 'conv1d', 'conv3d',
>             'linear', 'relu', 'sigmoid', 'tanh', 'softmax'
>         ]
>         
>         # Monkey patch torch functions
>         for op_name in ops_to_throttle:
>             if hasattr(torch, op_name):
>                 original_op = getattr(torch, op_name)
>                 self.original_ops[op_name] = original_op
>                 
>                 def make_throttled_op(orig_op):
>                     def throttled_op(*args, **kwargs):
>                         result = orig_op(*args, **kwargs)
>                         time.sleep(self.delay)
>                         return result
>                     return throttled_op
>                 
>                 setattr(torch, op_name, make_throttled_op(original_op))
>         
>         # Monkey patch torch.nn.functional
>         import torch.nn.functional as F
>         functional_ops = ['conv2d', 'linear', 'relu', 'softmax']
>         
>         for op_name in functional_ops:
>             if hasattr(F, op_name):
>                 original_op = getattr(F, op_name)
>                 self.original_ops[f'F.{op_name}'] = original_op
>                 
>                 def make_throttled_functional_op(orig_op):
>                     def throttled_op(*args, **kwargs):
>                         result = orig_op(*args, **kwargs)
>                         time.sleep(self.delay)
>                         return result
>                     return throttled_op
>                 
>                 setattr(F, op_name, make_throttled_functional_op(original_op))
>         
>         self.enabled = True
>     
>     def remove_patches(self):
>         """Remove all monkey patches"""
>         if not self.enabled:
>             return
>             
>         import torch.nn.functional as F
>         
>         for op_name, original_op in self.original_ops.items():
>             if op_name.startswith('F.'):
>                 setattr(F, op_name[2:], original_op)
>             else:
>                 setattr(torch, op_name, original_op)
>         
>         self.original_ops.clear()
>         self.enabled = False
> 
> # Usage
> throttler = GPUOpThrottler(delay=0.005)
> throttler.throttle_tensor_ops()
> You can increase/decrease the delay for more/less throttling.

Interesting, I'll look at it.

My program has two threads running for frame capture and frame processing / display, should i run "throttler.throttle_tensor_ops()"  once before the while loop in my Display thread starts, or does it need to be called before every Yolo prediction?

**Comment by Y-T-G** [comment]
It needs to be run once

**Comment by MichaelBrambleberry** [comment]
One last question, is there an equivalent for plain YOLOv11n? Your code seems to work with tensor, I was using tensor but I realized it won't be suitable for my purposes because the program is meant to run on Windows, and Tensor stopped supporting GPU acceleration with windows a few versions ago.

**Comment by Y-T-G** [comment]
It's for the PyTorch model (`.pt`). If you're using the `.pt` model, then it should work.
